aggregator/operator_registrations_inmemory.go

```diff
@@ -0,0 +1,190 @@
+// It takes a context as argument because the "backfilling" of the database is done inside this constructor,
+// so we wait for all past NewPubkeyRegistration events to be queried and the db to be filled before returning the service.
+// The constructor is thus following a RAII-like pattern, of initializing the serving during construction.
+// Using a separate initialize() function might lead to some users forgetting to call it and the service not behaving properly.
+func NewOperatorRegistrationsServiceInMemory(
+	ctx context.Context,
+	avsRegistrySubscriber avsregistry.AvsRegistrySubscriber,
+	avsRegistryReader avsregistry.AvsRegistryReader,
+	logger logging.Logger,
+) *OperatorRegistrationsServiceInMemory {
+	queryByAddrC := make(chan queryByAddr)
+	queryByIdC := make(chan queryById)
+
+	pkcs := &OperatorRegistrationsServiceInMemory{
+		avsRegistrySubscriber: avsRegistrySubscriber,
+		avsRegistryReader:     avsRegistryReader,
+		logger:                logger,
+		queryByAddrC:          queryByAddrC,
+		queryByIdC:            queryByIdC,
+	}
+
+	// We use this waitgroup to wait on the initialization of the inmemory pubkey dict,
+	// which requires querying the past events of the pubkey registration contract
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	pkcs.startServiceInGoroutine(ctx, queryByAddrC, queryByIdC, &wg)
+	wg.Wait()
+
+	return pkcs
+}
+
+func (ors *OperatorRegistrationsServiceInMemory) startServiceInGoroutine(ctx context.Context, queryByAddrC <-chan queryByAddr, queryByIdC <-chan queryById, wg *sync.WaitGroup) {
+	go func() {
+		pubkeyByAddrDict := make(map[common.Address]types.OperatorPubkeys)
+		pubkeyByIdDict := make(map[types.OperatorId]types.OperatorPubkeys)
+
+		ors.logger.Debug("Subscribing to new pubkey registration events on blsApkRegistry contract", "service", "OperatorRegistrationsServiceInMemory")
+		newPubkeyRegistrationC, newPubkeyRegistrationSub, err := ors.avsRegistrySubscriber.SubscribeToNewPubkeyRegistrations()
+		if err != nil {
+			ors.logger.Error("Fatal error opening websocket subscription for new pubkey registrations", "err", err, "service", "OperatorRegistrationsServiceInMemory")
+			// see the warning above the struct definition to understand why we panic here
+			panic(err)
+		}
+
+		ors.queryPastRegisteredOperatorEventsAndFillDb(ctx, pubkeyByAddrDict, pubkeyByIdDict)
+
+		// The constructor can return after we have backfilled the db by querying the events of operators that have registered with the blsApkRegistry
+		// before the block at which we started the ws subscription above
+		wg.Done()
+
+		for {
+			select {
+			case <-ctx.Done():
+				ors.logger.Infof("OperatorRegistrationsServiceInMemory: Context cancelled, exiting")
+				return
+
+			case err := <-newPubkeyRegistrationSub.Err():
+				ors.logger.Error("Error in websocket subscription for new pubkey registration events. Attempting to reconnect...", "err", err, "service", "OperatorRegistrationsServiceInMemory")
```

Comment:
We can just give up here if it's a SafeClient, no need to worry about resubs

Reply:
Giving up now, though the service still works responds unitil ctx.Done